lib/mix/lib/mix/tasks/app.start.ex

```diff
@@ -108,15 +108,33 @@ defmodule Mix.Tasks.App.Start do
   end
 
   defp ensure_all_started(app, type) do
-    case Application.ensure_all_started(app, type) do
-      {:ok, _} ->
+    case Application.start(app, type) do
+      :ok ->
         :ok
 
-      {:error, {app, reason}} ->
-        Mix.raise("Could not start application #{app}: " <> Application.format_error(reason))
+      {:error, {:already_started, ^app}} ->
+        :ok
+
+      {:error, {:not_started, dep}} ->
+        :ok = ensure_all_started(dep, type)
+        ensure_all_started(app, type)
+
+      {:error, reason} when type == :permanent ->
+        # We need to stop immediately because application_controller is
+        # shutting down bringing the system down with it. Any work here
+        # is prone to race conditions.
+        Mix.shell().error(["** (Mix) ", could_not_start(app, reason)])
+        :init.stop(1)
+
+      {:error, reason} ->
+        Mix.raise(could_not_start(app, reason))
```

Comment:
Compared to today, this implementation does not stop the applications we have already started. I don't think this is a problem because we start applications at different moments in Mix already, so you always got an inconsistent state.